README.md
[![Build Status](https://travis-ci.org/MSeal/cython_hunspell.svg?branch=master)](https://travis-ci.org/MSeal/cython_hunspell)

# CyHunspell
Cython wrapper on Hunspell Dictionary

## Description
This repository provides a wrapper on Hunspell to be used natively in Python. The
module uses cython to link between the C++ and Python code, with some additional
features. There's very little Python overhead as all the heavy lifting is done
on the C++ side of the module interface, which gives optimal performance.

The hunspell library will cache any corrections, you can use persistent caching by
adding the `use_disk_cache` argument to a Hunspell constructor. Otherwise it uses
in-memory caching.

## Dependencies
cacheman -- for (optionally asynchronous) persistent caching

## Non-Python Dependencies

### hunspell
If you don't have hunspell installed the library will download it for you.

If you want to use ubuntu's 1.6 installation use the following before installing:

    sudo apt-get install libhunspell-1.6-0 libhunspell-dev

This is a faster install than the built-in download and compile.

## Features
Spell checking & spell suggestions
* See http://hunspell.sourceforge.net/

## How to use
Below are some simple examples for how to use the repository.

### Creating a Hunspell object
    from hunspell import Hunspell
    h = Hunspell()

You now have a usable hunspell object that can make basic queries for you.

    h.spell('test') # True

### Spelling
It's a simple task to ask if a particular word is in the dictionary.

    h.spell('correct') # True
    h.spell('incorect') # False

This will only ever return True or False, and won't give suggestions about why it
might be wrong. It also depends on your choice of dictionary.

### Suggestions
If you want to get a suggestion from Hunspell, it can provide a corrected label
given a basestring input.

    h.suggest('incorect') # (u'incorrect', u'correction', u'corrector', u'correct', u'injector')

The suggestions are in sorted order, where the lower the index the closer to the
input string.

### Stemming
The module can also stem words, providing the stems for pluralization and other
inflections.

    h.stem('testers') # (u'tester', u'test')
    h.stem('saves') # (u'save',)

### Bulk Requests
You can also request bulk actions against Hunspell. This will trigger a threaded
(without a gil) request to perform the action requested. Currently just 'suggest'
and 'stem' are bulk requestable.

    h.bulk_suggest(['correct', 'incorect'])
    # {'incorect': (u'incorrect', u'correction', u'corrector', u'correct', u'injector'), 'correct': ['correct']}
    h.bulk_stem(['stems', 'currencies'])
    # {'currencies': [u'currency'], 'stems': [u'stem']}

By default it spawns number of CPUs threads to perform the operation. You can
overwrite the concurrency as well.

    h.set_concurrency(4) # Four threads will now be used for bulk requests

### Dictionaries
You can also specify the language or dictionary you wish to use.

    h = Hunspell('en_CA') # Canadian English

By default you have the following dictionaries available
* en_AU
* en_CA
* en_GB
* en_NZ
* en_US
* en_ZA

However you can download your own and point Hunspell to your custom dictionaries.

    h = Hunspell('en_GB-large', hunspell_data_dir='/custom/dicts/dir')

### Asynchronous Caching
If you want to have Hunspell cache suggestions and stems you can pass it a directory
to house such caches.

    h = Hunspell(disk_cache_dir='/tmp/hunspell/cache/dir')

This will save all suggestion and stem requests periodically and in the background.
The cache will fork after a number of new requests over particular time ranges and
save the cache contents while the rest of the program continues onward. You'll never
have to explicitly save your caches to disk, but you can if you so choose.

    h.save_cache()

Otherwise the Hunspell object will cache such requests locally in memory and not
persist that memory.

## Platforms
### Linux
Tested on Ubuntu and Fedora with pre-build binaries of Hunspell as well as
automatically build depedencies. It's inlikely to have trouble with other
distributions.

### Windows
The base library comes with MSVC built Hunspell libraries and will link
against those during runtime. These were tested on Windows 7, 8, 10 and
some on older systems. It's possible that a Python build with a newer
(or much older) version of MSVC will fail to load these pre-built libraries.

### Mac OSX
So far the library has been tested against 10.9 (Mavericks) and up. There
shoudn't be any reason it would fail to run on any particular version of
OSX.

## Building source libraries
See libs/README

## Navigating the Repo
### hunspell
Package wrapper for the repo.

### tests
All unit tests for the repo.

## Language Preferences
* Google Style Guide
* Object Oriented (with a few exceptions)

## TODO
* Convert cacheman dependency to be optional

## Known Issues
- Exact spelling suggestions on different OS's differs slightly with identical
inputs. This appears to be an issue with Hunspell 1.3 and not this library.
- Older versions of pip and setuptools will build with incorrect windows DLL bindings
and complain about "ImportError: DLL load failed: %1 is not a valid Win32 application."
- Sometimes windows machines won't find the build tools appropiately. You may need
to 'SET VS100COMNTOOLS=%VSxxxCOMNTOOLS%' before installing. Python 3 usually wants the
xxx as '140' and python 2 as '90'. There's not a lot the library can do to fix this,
though pip and setuptools upgrades oftentimes resolve the issue by being smarter.

## Author
Author(s): Tim Rodriguez and Matthew Seal

## License
MIT


CHANGELOG.md
# 1.3.0
- Fixed build for python 3.7
- Fixed library search issues (> Ubunutu 17)
- Upgraded default hunspell to 1.6.2 for Linux distros

# 1.2.1
- Fixed empty string crash

# 1.2.0
- Fixed detect CPU issue on Linux distros
- Fixed bytes versus unicode conversion for inputs in python2
- Added fix for Python 2.7 on osx
- Added fix for Windows 10 builds

# 1.1.4
- Added Python 3 support

# 1.1.3
- Removed library depdency on cython
- Dropped support for Python 2.6
- Added ability to set concurrency on bulk operations


tests/parentpath.py
import sys
import os

# Add parent import capabilities
parentdir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
if parentdir not in sys.path:
    sys.path.insert(0, parentdir)


tests/hunspell_test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# This import fixes sys.path issues
from . import parentpath

import os
import unittest
from hunspell import Hunspell

DICT_DIR = os.path.join(os.path.dirname(__file__), '..', 'dictionaries')

class HunspellTest(unittest.TestCase):
    def setUp(self):
        self.h = Hunspell('test', hunspell_data_dir=DICT_DIR)

    def tearDown(self):
        try:
            del self.h
        except AttributeError:
            pass

    def assertAllIn(self, checked, expected):
        self.assertTrue(all(x in expected for x in checked),
            u"{} not all found in {}".format(checked, expected))

    def test_create_destroy(self):
        del self.h

    def test_missing_dict(self):
        with self.assertRaises(IOError):
            Hunspell('not_avail', hunspell_data_dir=DICT_DIR)

    def test_spell(self):
        self.assertFalse(self.h.spell('dpg'))
        self.assertTrue(self.h.spell('dog'))

    def test_spell_utf8(self):
        self.assertTrue(self.h.spell(u'café'))
        self.assertFalse(self.h.spell(u'uncafé'))

    def test_spell_empty(self):
        self.assertTrue(self.h.spell(''))

    def test_suggest(self):
        required = ('dog', 'pg', 'deg', 'dig', 'dpt', 'dug', 'mpg', 'd pg')
        suggest = self.h.suggest('dpg')
        self.assertIsInstance(suggest, tuple)
        self.assertAllIn(required, suggest)

    def test_suggest_utf8(self):
        required = (u'café', u'Cerf')
        for variant in ('cefé', u'cefé'):
            suggest = self.h.suggest(variant)
            self.assertIsInstance(suggest, tuple)
            self.assertAllIn(required, suggest)

    def test_suggest_empty(self):
        self.assertEqual(self.h.suggest(''), ())

    def test_stem(self):
        self.assertEqual(self.h.stem('dog'), ('dog',))
        self.assertEqual(self.h.stem('permanently'), ('permanent',))

    def test_bulk_suggest(self):
        self.h.set_concurrency(3)
        suggest = self.h.bulk_suggest(['dog', 'dpg'])
        self.assertEqual(sorted(suggest.keys()), ['dog', 'dpg'])
        self.assertIsInstance(suggest['dog'], tuple)
        self.assertAllIn(('dog',), suggest['dog'])

        required = ('dog', 'pg', 'deg', 'dig', 'dpt', 'dug', 'mpg', 'd pg')
        self.assertIsInstance(suggest['dpg'], tuple)
        self.assertAllIn(required, suggest['dpg'])

        checked = ['bjn', 'dog', 'dpg', 'dyg', 'foo', 'frg', 'opg', 'pgg', 'qre', 'twg']
        suggest = self.h.bulk_suggest(checked)
        self.assertEqual(sorted(suggest.keys()), checked)

    def test_bulk_stem(self):
        self.h.set_concurrency(3)
        self.assertDictEqual(self.h.bulk_stem(['dog', 'permanently']), {
            'permanently': ('permanent',),
            'dog': ('dog',)
        })
        self.assertDictEqual(self.h.bulk_stem(['dog', 'twigs', 'permanently', 'unrecorded']), {
            'unrecorded': ('recorded',),
            'permanently': ('permanent',),
            'twigs': ('twig',),
            'dog': ('dog',)
        })

if __name__ == '__main__':
    unittest.main()


tar_download.py
import os
import sys
import tarfile
try:
    from urllib import urlretrieve
except ImportError:
    from urllib.request import urlretrieve

def file_name_from_url(url, directory=None):
    file_name = url.split('/')[-1]
    if directory:
        file_name = os.path.join(directory, file_name)
    return file_name

def check_file_exists(url, directory=None, skipFound=True):
    file_name = file_name_from_url(url, directory)
    if os.path.exists(file_name):
        if skipFound:
            print("File already found: {}, skipping download of {}".format(
                file_name, url))
        return True
    return False

def download_tar(url, directory=None, skipFound=True):
    if skipFound and check_file_exists(url, directory, skipFound):
        return

    if not os.path.exists(directory):
        os.makedirs(directory)

    file_name = file_name_from_url(url, directory)
    print("Downloading {} to {}".format(url, file_name))
    sys.stdout.flush()
    urlretrieve(url, file_name)

def extract_contents(file_name, destination='.'):
    if not os.path.exists(destination):
        os.makedirs(destination)

    print("Extracting {} to {}".format(file_name, destination))
    sys.stdout.flush()
    tar = tarfile.open(file_name)
    tar.extractall(destination)
    tar.close()

def download_and_extract(url, directory=None, skipFound=True):
    if skipFound and check_file_exists(url, directory, skipFound):
        return

    download_tar(url, directory)
    extract_contents(file_name_from_url(url, directory), directory)


hunspell/thread.hpp
#ifndef AGNOSTIC_THREAD_HPP_
#define AGNOSTIC_THREAD_HPP_

typedef void * thread_t;

#ifdef _MSC_VER
#include <windows.h>
void dealloc_threads(thread_t **threads, int num_threads) {
    if (threads != NULL) {
        for (int i = 0; i < num_threads; i++) {
            if (threads[i] != NULL) {
                CloseHandle(threads[i]);
            }
        }
        free(threads);
    }
}

struct WorkerWrapperArgs {
    void *(*worker)(void *);
    void *data;
};

unsigned long __stdcall worker_wrapper(void *args) {
    void *data = ((WorkerWrapperArgs *)args)->data;
    void *(*worker)(void *) = ((WorkerWrapperArgs *)args)->worker;
    free(args);

    void *worker_result = worker(data);
    return worker_result == NULL ? 0 : 1;
}

thread_t *thread_create(void *(*worker)(void *), void *data) {
    unsigned long id_holder;
    WorkerWrapperArgs *args = (WorkerWrapperArgs *)malloc(sizeof(WorkerWrapperArgs));
    args->worker = worker;
    args->data = data;
    return (thread_t *)CreateThread(NULL, 0, &worker_wrapper, args, 0, &id_holder);
}

int thread_join(thread_t *thread) {
    return WaitForSingleObject(thread, INFINITE);
}
#else
#include <pthread.h>
void dealloc_threads(thread_t **threads, int num_threads) {
    if (threads != NULL) {
        for (int i = 0; i < num_threads; i++) {
            if (threads[i] != NULL) {
                free(threads[i]);
            }
        }
        free(threads);
    }
}

thread_t *thread_create(void *(*worker)(void *), void *data) {
    pthread_t *thread = (pthread_t *)malloc(sizeof(pthread_t));
    if (pthread_create(thread, NULL, worker, data)) {
        free(thread);
        thread = NULL;
    }
    return (thread_t *)thread;
}

int thread_join(thread_t *thread) {
    pthread_t *pt = (pthread_t *)thread;
    return pthread_join(*pt, NULL);
}
#endif

#endif /* AGNOSTIC_THREAD_HPP_ */


find_library.py
#
# This is a cross-platform shared library search mechanism. Tested on:
# OSX, Ubuntu, Fedora, and Windows XP+
#
import os
import glob
import platform
import re
import sys
import shutil
from subprocess import check_call
from tar_download import download_and_extract
from distutils.sysconfig import get_python_lib
try:
    from subprocess import getstatusoutput
except ImportError:
    from commands import getstatusoutput

BASE_DIR = os.path.abspath(os.path.dirname(__file__))

def get_architecture():
    return 'x64' if sys.maxsize > 2**32 else 'x86'

def get_prefered_msvc():
    if sys.version_info[0] > 3 or sys.version_info[0] < 2:
        raise RuntimeError('Unknown python version')
    # Python 2.6-3.2
    if sys.version_info[0] < 3 or sys.version_info[1] < 3:
        # We should return this, but the 11 build works and we provide it
        #return 'msvc9'
        return 'msvc11'
    # Python 3.3-3.4
    elif sys.version_info[1] < 5:
        # We should return this, but the 11 build works and we provide it
        #return 'msvc10'
        return 'msvc11'
    # Python 3.5+
    else:
        # These versions need msvc14+ compilations
        return 'msvc14'

def form_possible_names(lib, exts, extact=False):
    ret = []
    for ext in exts:
        if not extact:
            ret.append('{}*{}'.format(lib, ext))
        else:
            ret.append('{}{}'.format(lib, ext))
    for ext in exts:
        if not extact:
            ret.append('lib{}*{}'.format(lib, ext))
        else:
            ret.append('lib{}{}'.format(lib, ext))
    return ret

def do_search(paths, names=[], test_fn=None):
    for pn in paths:
        globbed = []
        for name in names:
            if '*' in name:
                globbed.extend(glob.glob(os.path.join(pn, name)))
            else:
                globbed.append(os.path.join(pn, name))
        for filepath in globbed:
            if test_fn:
                if test_fn(filepath):
                    return filepath, pn
            elif os.path.exists(filepath):
                return filepath, pn
    return None, None

def is_library(filepath, acceptable_exts):
    # TODO - This is broken for ".dll.a"
    return os.path.isfile(filepath) and (os.path.splitext(filepath)[-1] in acceptable_exts)

def is_header(filepath):
    return os.path.isfile(filepath)

def include_dirs(*packages):
    dirs = []
    if 'hunspell' in packages:
        if platform.system() == 'Linux':
            dirs = [
                os.path.abspath(os.path.join(BASE_DIR, 'hunspell')),
                # Download path if missing
                os.path.abspath(os.path.join(BASE_DIR, 'external', 'hunspell-1.6.2', 'src')),
            ]
        else:
            dirs = [
                os.path.abspath(os.path.join(BASE_DIR, 'hunspell')),
                # Download path if missing
                os.path.abspath(os.path.join(BASE_DIR, 'external', 'hunspell-1.3.3', 'src')),
            ]
    if platform.system() != 'Windows':
        dirs.extend([
            '/usr/local/include',
            '/opt/include',
            '/usr/include'
        ])
    return [path for path in dirs if os.path.isdir(path)]

def library_dirs(check_local=False):
    dirs = [os.path.abspath(BASE_DIR)]
    if platform.system() == 'Windows':
        dirs.extend([
            os.path.dirname(__file__),
            os.path.abspath(BASE_DIR),
            os.path.join(os.environ.get('SystemRoot'), 'system'),
            os.path.join(os.environ.get('SystemRoot'), 'system32'),
            os.environ.get('SystemRoot')
        ])
        if check_local:
            dirs.append(os.path.join(os.path.dirname(__file__), 'libs', 'msvc'))
        dirs.extend(list(set(os.environ.get('PATH').split(os.path.pathsep))))
        dirs = [os.path.abspath(path) for path in dirs]
    else:
        dirs.extend([
            os.path.abspath(os.path.join(get_python_lib(), 'libs', 'unix')),
            '/usr/local/lib64',
            '/usr/local/lib',
            '/usr/local/libdata',
            '/opt/local/lib',
            '/usr/lib/x86_64-linux-gnu',
            '/usr/lib64',
            '/usr/lib',
            '/usr/X11/lib',
            '/usr/share',
        ])
    dirs = list(set(dirs))
    try:
        while True:
            dirs.remove(None)
    except ValueError:
        pass
    return [path for path in dirs if os.path.isdir(path)]

def get_library_path(lib, check_local=False):
    paths = library_dirs(check_local)
    acceptable_exts = [
        '',
        '.so'
    ]

    if platform.system() == 'Windows':
        acceptable_exts = [
            '.lib'
        ]
    elif platform.system() == 'Darwin':
        acceptable_exts.append('.dylib')

    names = form_possible_names(lib, acceptable_exts, check_local)
    found_lib, found_path = do_search(paths, names, lambda filepath: is_library(filepath, acceptable_exts))
    if found_lib and platform.system() == 'Windows':
        found_lib = os.path.splitext(found_lib)[0]
    return found_lib, found_path

def get_library_linker_name(lib):
    found_lib, found_path = get_library_path(lib)
    if not found_lib:
        # Try x86 or x64
        found_lib, found_path = get_library_path(lib + get_architecture(), True)
        if not found_lib:
            found_lib, found_path = get_library_path('-'.join(
                [lib, get_prefered_msvc(), get_architecture()]), True)

    if found_lib:
        found_lib = re.sub(r'.dylib$|.so$|.dll$|.dll.a$|.a$', '', found_lib.split(os.path.sep)[-1])
        if platform.system() != 'Windows':
            found_lib = re.sub(r'^lib|', '', found_lib)

    return found_lib, found_path

def package_found(package, include_dirs):
    for idir in include_dirs:
        package_path = os.path.join(idir, package)
        if os.path.exists(package_path) and os.access(package_path, os.R_OK):
            return True
    return False

def build_hunspell_package(directory, force_build=False):
    tmp_lib_path = os.path.abspath(os.path.join(os.path.dirname(__file__), 'libs', 'tmp'))
    if not os.path.exists(tmp_lib_path):
        os.makedirs(tmp_lib_path)

    olddir = os.getcwd()
    if force_build or not os.path.exists(os.path.join(tmp_lib_path, 'lib', 'libhunspell-1.6.so.0.0.1')):
        try:
            os.chdir(directory)
            check_call(['autoreconf', '-vfi'])
            check_call(['./configure', '--prefix='+tmp_lib_path])
            check_call('make')
            check_call(['make', 'install'])
        finally:
            os.chdir(olddir)

    lib_path = os.path.abspath(os.path.join(get_python_lib(), 'libs', 'unix'))
    if os.path.exists(lib_path):
        shutil.rmtree(lib_path)
    os.makedirs(lib_path)

    shutil.copyfile(
        os.path.join(tmp_lib_path, 'lib', 'libhunspell-1.6.so.0.0.1'),
        os.path.join(lib_path, 'libhunspell-1.6.so.0'))
    os.symlink(
        os.path.join(lib_path, 'libhunspell-1.6.so.0'),
        os.path.join(lib_path, 'libhunspell.so'))
    shutil.rmtree(tmp_lib_path)

    return lib_path

def append_links(pkg, kw):
    linker_name, linker_path = get_library_linker_name(pkg)
    if linker_name:
        kw['libraries'].append(linker_name)
    if linker_path:
        kw['library_dirs'].append(linker_path)
    if linker_path and platform.system() != 'Windows':
        kw['runtime_library_dirs'].append(linker_path)
    return linker_name

def pkgconfig(*packages, **kw):
    try:
        flag_map = {'-I': 'include_dirs', '-L': 'library_dirs', '-l': 'libraries'}
        status, response = getstatusoutput("pkg-config --libs --cflags {}".format(' '.join(packages)))
        if status != 0:
            raise RuntimeError(response)
        for token in response.split():
            kw.setdefault(flag_map.get(token[:2]), []).append(token[2:])
            if token[:2] in flag_map:
                arg = flag_map.get(token[:2])
                kw.setdefault(arg, []).append(token[2:])
                kw[arg] = list(set(kw[arg]))
            else: # throw others to extra_link_args
                kw.setdefault('extra_link_args', []).append(token)
                kw['extra_link_args'] = list(set(kw['extra_link_args']))
    except RuntimeError:
        kw['include_dirs'] = include_dirs(*packages)
        kw['library_dirs'] = []
        kw['runtime_library_dirs'] = []
        kw['libraries'] = []
        kw['extra_link_args'] = []

        if 'hunspell' in packages and not package_found('hunspell', kw['include_dirs']):
            # Prepare for hunspell if it's missing
            if not os.environ.get('SKIP_DOWNLOAD', False):
                if platform.system() == 'Linux':
                    download_and_extract('https://github.com/hunspell/hunspell/archive/v1.6.2.tar.gz',
                        os.path.join(BASE_DIR, 'external'))
                else:
                    download_and_extract('http://downloads.sourceforge.net/hunspell/hunspell-1.3.3.tar.gz',
                        os.path.join(BASE_DIR, 'external'))
                kw['include_dirs'] = include_dirs(*packages)
            else:
                raise RuntimeError("Could not find hunspell and not allowed to download")

        for pkg in packages:
            if not append_links(pkg, kw):
                if pkg == 'hunspell' and platform.system() == 'Linux':
                    lib_path = build_hunspell_package(os.path.join(BASE_DIR, 'external', 'hunspell-1.6.2'))
                    if not append_links(pkg, kw):
                        raise RuntimeError("Couldn't find lib dependency after building: {}".format(pkg))
                    else:
                        kw['extra_link_args'] += ['-Wl,-rpath,"{}"'.format(lib_path)]
                elif pkg == 'hunspell' and platform.system() != 'Windows':
                    lib_path = build_hunspell_package(os.path.join(BASE_DIR, 'external', 'hunspell-1.3.3'))
                    if not append_links(pkg, kw):
                        raise RuntimeError("Couldn't find lib dependency after building: {}".format(pkg))
                    else:
                        kw['extra_link_args'] += ['-Wl,-rpath,"{}"'.format(lib_path)]
                else:
                    raise RuntimeError("Couldn't find lib dependency: {}".format(pkg))

    return kw
